relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+///
+/// The collector is cheap to clone.
+#[derive(Clone)]
+pub struct Cogs {
+    recorder: Arc<dyn CogsRecorder>,
+}
+
+impl Cogs {
+    /// Creates a new [`Cogs`] from a [`recorder`](CogsRecorder).
+    pub fn new<T>(recorder: T) -> Self
+    where
+        T: CogsRecorder + 'static,
+    {
+        Self {
+            recorder: Arc::new(recorder),
+        }
+    }
+
+    /// Shortcut for creating a [`Cogs`] from a [`crate::NoopRecorder`].
+    ///
+    /// All collected measurements will be dropped.
+    pub fn noop() -> Self {
+        Self::new(crate::NoopRecorder)
+    }
+}
+
+impl Cogs {
+    /// Starts a recording for a COGS measurement.
+    ///
+    /// When the returned token is dropped the measurement will be recorded
+    /// with the configured [recorder](CogsRecorder).
+    ///
+    /// The recorded measurement can be attributed to multiple features by supplying a
+    /// weighted [`AppFeatures`]. A single [`AppFeature`] attributes the entire measurement
+    /// to the feature.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// # use relay_cogs::{AppFeature, Cogs, ResourceId};
+    /// # struct Span;
+    /// # fn scrub_sql(_: &mut Span) {}
+    /// # fn extract_tags(_: &mut Span) {};
+    ///
+    /// fn process_span(cogs: &Cogs, span: &mut Span) {
+    ///     let _token = cogs.timed(ResourceId::Relay, AppFeature::Spans);
+    ///
+    ///     scrub_sql(span);
+    ///     extract_tags(span);
+    /// }
+    ///
+    /// ```
+    pub fn timed<F: Into<AppFeatures>>(&self, resource: ResourceId, features: F) -> CogsToken {
+        CogsToken {
+            resource,
+            features: features.into(),
+            start: Instant::now(),
+            recorder: Arc::clone(&self.recorder),
+        }
+    }
+}
+
+/// An in progress COGS measurement.
+///
+/// The measurement is recorded when the token is dropped.
+#[must_use]
+pub struct CogsToken {
+    resource: ResourceId,
+    features: AppFeatures,
+    start: Instant,
+    recorder: Arc<dyn CogsRecorder>,
+}
+
+impl CogsToken {
+    /// Cancels the COGS measurement.
+    pub fn cancel(&mut self) {
+        // No features -> nothing gets attributed.
+        self.update(AppFeatures::none());
+    }
+
+    /// Updates the app features to whicht the active measurement is attributed to.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// # use relay_cogs::{AppFeature, Cogs, ResourceId};
+    /// # struct Item;
+    /// # fn do_something(_: &Item) -> bool { true };
+    ///
+    /// fn process(cogs: &Cogs, item: &Item) {
+    ///     let mut token = cogs.timed(ResourceId::Relay, AppFeature::Unattributed);
+    ///
+    ///     // App feature is only known after some computations.
+    ///     if do_something(item) {
+    ///         token.update(AppFeature::Spans);
+    ///     } else {
+    ///         token.update(AppFeature::Transactions);
+    ///     }
+    /// }
+    /// ```
+    pub fn update<T: Into<AppFeatures>>(&mut self, features: T) {
```

Comment:
Instead of having an interface that overwrites all previous app features, could we have an interface that allows multiple updates? I.e.

```rust
let mut accountant = CogsToken::new();
// ...
for bucket in buckets {
    let feature = AppFeature::from(bucket.namespace);
    accountant.attribute(feature, bucket.size());
}
// in drop handler, accountant divides up elapsed time proportional to attributed numbers.
```

Reply:
Hmm maybe I should keep on reading, seems like the builder does what I had in mind.

Reply:
The update was a bit of an afterthought, I was hoping I can set the weights once, when starting the measurement. Turns out we have some edge cases where that is not possible and we first have to do work before we can attribute the measurement.